[JENKINS-27177] Add disable plugin command and tests

MRamonLeon opened the pull request:
See [JENKINS-27177](https://issues.jenkins-ci.org/browse/JENKINS-27177).

Added a new CLI command `disable-plugin` to disable one or more plugins and optionally restart Jenkins.

Behavior:
* The info about the plugins successfully disabled are shown in the output console.
* If a plugin is already disabled, it show it in the output and exit normally (error code 0).
* If a plugin in the list can't be disabled, the process continues up to the end of the list of plugins passed and the command exits with error code 16 and the message about this plugin is output in the error console.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* [JENKINS-27177](https://issues.jenkins-ci.org/browse/JENKINS-27177): Add a new CLI command `disable-plugin` to disable one ore more installed plugins and optionally restart Jenkins.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @jvz 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

![screenshot from 2018-09-25 14-54-24](https://user-images.githubusercontent.com/22069922/46015694-371bc780-c0d3-11e8-9737-7a4438c7c3ec.png)

If you call it with:
`java -jar jenkins-cli.jar ... disable-plugin variant dependee mandatory-depender depender`

It responds with:

```
The plugin variant was already disabled.
The plugin dependee has, at least, one dependant plugin (mandatory-depender), so it cannot be disabled.
Plugin mandatory-depender disabled.
Plugin depender disabled.
```
Return code: 16 (because one plugin couldn't be disabled)

MRamonLeon committed:
[JENKINS-27177] Add disable plugin command and tests

fcojfernandez commented on core/src/main/resources/hudson/cli/Messages.properties:
```diff
@@ -113,3 +113,12 @@ WaitNodeOfflineCommand.ShortDescription=Wait for a node to become offline.
 CliProtocol.displayName=Jenkins CLI Protocol/1 (unencrypted)
 CliProtocol2.displayName=Jenkins CLI Protocol/2 (transport encryption)
+
+DisablePluginCommand.ShortDescription=\
```
nit: why the `\`?

MRamonLeon replied:
To introduce a new line, used in many other `ShortDescription` messages.

Wadeck replied:
Not nit for me, just a typo that should be correct 👍

Wadeck replied:
Initially I also put a negative comment on that but if you look at the file, it's full of this kind of stuff. I imagine that compared to HTML our CLI is not eating the initial blank lines ;)

fcojfernandez replied:
Indeed. It's just a matter of taste

fcojfernandez commented on test/src/test/java/hudson/cli/DisablePluginCommandTest.java:
```diff
@@ -0,0 +1,170 @@
+    public JenkinsRule j = new JenkinsRule();
+
+    @Test
+    @Issue("JENKINS-27177")
```
I think the @Issue is used for bugs only

MRamonLeon replied:
It's used in the `enable-plugin` too. It's not a bug. That's why I used it too.

Wadeck replied:
Initially the JavaDoc says it's for "bug filed in the issue tracker" but now we use it for non-regression testing as well.

fcojfernandez commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+ * @since 2.136
+ */
+@Extension
+public class DisablePluginCommand extends CLICommand {
```
How would it be invoked? A screenshot of the help window would be useful.

MRamonLeon replied:
done

fcojfernandez commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        if (restart && somePluginDisabled) {
+            jenkins.safeRestart();
```
Is that an asynchronous execution? I mean, will the CLI wait for the restart?

MRamonLeon replied:
It doesn't wait. The restart will begin when there aren't jobs running

fcojfernandez commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+            switch (disablePlugin(manager, pluginName)) {
+                case DISABLED:
+                    if(!somePluginDisabled) somePluginDisabled = true;
+                    break;
```
Why the break? Don't we want to disable all the plugins?

MRamonLeon replied:
the break is for the switch, but the loop continues.

Wadeck replied:
The `break` is only for the related `case`, so seems fine.

fcojfernandez commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+    private DISABLING_STATUS disablePlugin(PluginManager manager, String shortName) throws IOException {
+        PluginWrapper plugin = manager.getPlugin(shortName);
+        if (plugin == null) {
+            throw new IllegalArgumentException(Messages.DisablePluginCommand_NoSuchPlugin(shortName)); // exit with 3
```
Is not catching the exception above on purpose?

MRamonLeon replied:
Indeed. As other commands, if the parameters are wrong, the process exists with the error code 3 thanks to this specific exception type.

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+ * continues, but the exit code is 16, instead of 0. Disabling an already disabled plugin does nothing. It only restart
+ * if, at least, one plugin has been disabled and the restart option was set.
+ *
+ * @since 2.136
```
Please write `@since TODO` as you may not know right now the version it will be published in. 
_(in addition with that we could too easily guessed from which class you copied your base code)_

MRamonLeon replied:
haha, I never tried to hide it

amuniz commented:
I think this CLI will need a more complex implementation to work properly.

Let's say plugin C depends on B and B depends on A. If you pass `C, B, A` the command (as written now) will do its job, however `A, B, C` won't work.

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+public class DisablePluginCommand extends CLICommand {
+
+    @Argument(metaVar = "plugins", required = true, usage = "Plugins to be disabled.")
+    private List<String> pluginNames;
```
TBH I dunno why, but the attributes are either let as public or put as private in general. Seems more natural to mark them as private.

MRamonLeon commented:
> I think this CLI will need a more complex implementation to work properly.

I think so too. But I wanted to have a first implementation to discuss a more complex one, messages and behavior.

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    protected void printUsageSummary(PrintStream stderr) {
+        stderr.println(Messages.DisablePluginCommand_PrintUsageSummary());
```
For `printUsageSummary` if you could follow the same format as in `LoginCommand` and `LogoutCommand` for consistency.

MRamonLeon replied:
I used ConsoleCommand, but no problem

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+    /**
+     * An enum to hold the status of a disabling action. To do it more reader-friendly.
+     */
+    private enum DISABLING_STATUS {
```
Please follow Java convention, even enum must be CamelCased => `DisablingStatus`.

MRamonLeon replied:
Right! it slipped my mind

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+                    if(!somePluginDisabled) somePluginDisabled = true;
+                    break;
+                case NOT_DISABLED_DEPENDANTS:
+                    if(result != 16) result = 16; // Custom error code to indicate one plugin, at least, can't be disabled
```
FTR Such integer code must be refactored to help readability. I have that item on my OSS todo list.

MRamonLeon replied:
I will appreciate it

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+        for (String dependant : dependants) {
+            PluginWrapper dependantPlugin = manager.getPlugin(dependant);
+            if (dependantPlugin != null && dependantPlugin.isEnabled()) {
+                // TO-DO: stdout or stderr? the process continues but as the result code is not 0, it looks like stderr
```
Please use `TODO` instead of `TO-DO` for better support in IDEs.

MRamonLeon replied:
I'll remove the comment, It was there to point it out to all reviewers

Wadeck commented:
> How to deal with optional dependencies

Perhaps @Vlatombe could help IIRC you did some improvement on the Wizard on that area?

MRamonLeon committed:
[JENKINS-27177] Add check to avoid restart in windows and a message fix.

MRamonLeon committed:
[JENKINS-27177] Some polishment.

Wadeck commented on test/src/test/java/hudson/cli/DisablePluginCommandTest.java:
```diff
@@ -167,4 +171,8 @@ private void assertJenkinsInQuietMode() {
     }
+
+    private void assumeNotWindows() {
+        Assume.assumeFalse(System.getProperty("os.name").startsWith("Windows"));
```
`Functions.isWindows` perhaps?

Vlatombe commented:
My latest PR was related to classloading of optional dependencies (https://github.com/jenkinsci/jenkins/pull/3370). Not really related with Wizard.

MRamonLeon committed:
[JENKINS-27177] Change windows check.

MRamonLeon commented:
I'll improve the procedure to allow disabling all plugins, no matter the order

jvz reviewed (approved):
LGTM. I like the `@WithPlugin` bit to simplify things!

fcojfernandez commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+        PluginManager manager = jenkins.getPluginManager();
+        int result = 0;
+        boolean somePluginDisabled = false; // to ensure we restart when it needs to
+        for (String pluginName : pluginNames) {
```
I think the comment by @amuniz makes sense. Maybe we can order the list of plugins matching a topological criteria and iterate that sorted list instead of the one passed as parameter.

MRamonLeon replied:
yes, it's in progress

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+        for (String pluginName : pluginNames) {
+            switch (disablePlugin(manager, pluginName)) {
+                case DISABLED:
+                    if(!somePluginDisabled) somePluginDisabled = true;
```
Why checking if already set? I think setting it directly is fine.

MRamonLeon replied:
to avoid setting it without being necessary.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+                    if(!somePluginDisabled) somePluginDisabled = true;
+                    break;
+                case NOT_DISABLED_DEPENDANTS:
+                    if(result != 16) result = 16; // Custom error code to indicate one plugin, at least, can't be disabled
```
Ditto.

MRamonLeon replied:
Ditto.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+    private List<String> pluginNames;
+
+    @Option(name = "-restart", usage = "Restart Jenkins after disabling plugins.")
+    private boolean restart;
```
I assume it's false if not present?

MRamonLeon replied:
yes, the default value for boolean.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+    private DisablingStatus disablePlugin(PluginManager manager, String shortName) throws IOException {
+        PluginWrapper plugin = manager.getPlugin(shortName);
+        if (plugin == null) {
+            throw new IllegalArgumentException(Messages.DisablePluginCommand_NoSuchPlugin(shortName)); // exit with 3
```
Why not continue with other plugins in this case?

MRamonLeon replied:
it's an option, I like it, but it's how the _enable-plugin_ works. I didn't want to do in the other way.

oleg-nenashev replied:
+1. Better to fail so that the error is propagated

fcojfernandez replied:
My question here: Let's suppose we want to disable A, B, C. A is disabled correctly, B does not exist so it throws the error and C then still remains enabled.
I like the idea of throwing the error, but I think an "all or nothing" approach would it better. We might check the existence of all plugins as first step before disabling any of them.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        plugin.disable();
+        stdout.format(Messages.DisablePluginCommand_Plugin_Disabled(plugin.getShortName()));
```
I'm not sure about printing the messages here. I think it makes more sense having the disable logic here and then printing the messages anywhere else.

MRamonLeon replied:
Could you explain it please?

varyvol replied:
To me it makes more sense than this command would return the result of disabling the plugin, and the place where it is called react to that result and prints the corresponding message.

fcojfernandez replied:
Maybe you're proposing like a final summary? It could be good idea.

varyvol commented on test/src/test/java/hudson/cli/DisablePluginCommandTest.java:
```diff
@@ -0,0 +1,179 @@
+    public void disablePluginWithMandatoryDepender() {
+        assertThat(disablePlugins("-restart", "dependee"), failedWith(16));
+        assertPluginEnabled("dependee");
+        assertJenkinsNotInQuietMode();
```
What does this command check?

MRamonLeon replied:
Wether Jenkins is restarting

varyvol reviewed (commented):
I agree this needs a more complex logic so that the order of the plugins doesn't influence the result.

oleg-nenashev reviewed (approved):
Looks good assuming that @Wadeck addressed the magic return codes issue in a follow-up

daniel-beck commented:
Based on comment history this still needs the ordering fix discussed above?

MRamonLeon commented:
You are right @daniel-beck, I'll push the improved version on monday, with different strategies to disable dependents plugins. Stay tuned ;-)

MRamonLeon committed:
[JENKINS-27177] Add strategies to disable plugins. Move process to core

Move the process to core to be reusable, not only for the cli command.
Add optionalDependants list to PluginWrapper to process them easily.

Replace (depreacate) the disable method without checks by this one with
strategies.

Improve messages and tests

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -946,10 +895,20 @@ public synchronized void resolveDependantPlugins() {
+
+                        // already know possibleDependant depends on plugin, no need to continue with the rest of
+                        // dependencies. We continue with the next possibleDependant
+                        break;
```
Please avoid using `break` in general (except in `switch`) as a good practice, to avoid attracting C developers ;)

(same for `continue`)

MRamonLeon replied:
I'm a one-single-point exit lover, and never ever use break or continue, but in this case I realized that this way was more understandable and maintainable. Even knowing that, I did a deep search about the topic and no clear winner come to me. So I ended up using it, to avoid extra control flags in bucles and ifs. Let's see what the rest of the guys think about it to take an action. WYT?

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -25,47 +25,29 @@
-import hudson.model.ItemGroupMixIn;
-import hudson.model.UpdateCenter;
-import hudson.model.UpdateSite;
+import hudson.model.*;
```
Please configure your IDE to avoid using star import. We are only accepting them in tests for the different assert/matchers

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1829,6 +1788,34 @@ public MetadataCache createCache() {
+     * Disable a list of plugins using a strategy for its dependants plugins.
+     * @param strategy the strategy regarding how the dependant plugins are processed
+     * @param plugins the list of plugins
+     * @return the list of results for every plugin and its dependant plugins.
```
and *their* dependant plugins? (same for the first sentence)

MRamonLeon commented:
I've implemented several ways to process the dependant plugins letting the administrator to use the one that best fits his/her needs.
![screenshot from 2018-10-12 20-32-52](https://user-images.githubusercontent.com/22069922/46937722-4e126180-d062-11e8-880b-96436987dd87.png)

- **none**: if a plugin has enabled-mandatory dependant plugins, it cannot be disabled.
- **mandatory**: all enabled-mandatory dependant plugins are disabled too.
- **all**: all enabled dependant plugins (mandatory or optional) are disabled too.

The process is moved to the core itself, allowing it to be used along the code, not only by the CLI command.

The former **disable** method (without any check) is deprecated in favor of the new one. 

Added the _optionalDependants_ list to the _PluginWrapper_ to be able to process it efficiently.

Improved the _PluginManager.resolveDependantPlugins_ method to avoid unnecessary iterations and added the _disablePlugins_ method.

In the command itself, I've renamed the option _cascade_ to _strategy_, added _quiet_ option, moved the procedure itself to the PluginManager and PluginWrapper classes to be able to reuse it and I keeped printing the output and calculating the result code in the command class.

Also improved messages and test class: add more tests and avoid having to know the output messages in the test class.

Wadeck commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -226,6 +232,14 @@ public void setDependants(@Nonnull Set<String> dependants) {
 
+    /**
+     * Get the list of components that depend optionally on this plugin.
+     * @return The list of components that depend optionally on this plugin.
```
nit: Not really useful to duplicate such JavaDoc, I would like to have only the `@return The list...` and the "global" JavaDoc does not add any information. (same for other methods)

Wadeck commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -519,9 +543,19 @@ public void enable() throws IOException {
+     * Disables this plugin next time Jenkins runs. As it doesn't check anything, it's recommended to use the method
+     * {@link #disable(PluginDisableStrategy)}
      */
+    @Deprecated //see https://issues.jenkins-ci.org/browse/JENKINS-27177
```
👍 for adding the reason of the deprecation!

Wadeck commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -530,6 +564,95 @@ public void disable() throws IOException {
+     * be raised because the disable file cannot be created.
+     */
+    @Nonnull
+    public PluginDisableResult disable(@Nonnull PluginDisableStrategy strategy) throws IOException {
```
nit: for the `@Nonnull` you put it once before the plugin, other time between public and the return typed. I would prefer the latter but at least keep consistency.

Wadeck commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -829,6 +952,86 @@ public static PluginWrapperAdministrativeMonitor get() {
+    /**
+     * An enum to hold the status of a disabling action against a plugin. To do it more reader-friendly.
+     */
+    public static enum PluginDisableStatus {
```
nit: `static` is redundant for inner enums.

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
+        // disable...
+        List<PluginWrapper.PluginDisableResult> results = jenkins.pluginManager.disablePlugins(strategyToUse, pluginNames);
+
+        // print results ...
```
nit: due to the good method naming, this comment is totally useless :)

MRamonLeon commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -78,84 +60,44 @@
-import org.kohsuke.stapler.HttpRedirect;
-import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.*;
```
Is it allowed? I don't know for sure

Wadeck replied:
It is not :) (except for assert/matchers in test)

MRamonLeon replied:
I had this doubt when I saw the IDE changed it

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
+        // print results ...
+        printResults(results);
+
+        // restart if it was required and it's necessary (some plugin was disabled in this execution)
```
some plugin was disabled in this execution => some plugin**s** w**ere** disabled in this execution

MRamonLeon replied:
It's intentionally singular because you just need one plugin disabled in this execution to raise the restart, I will change it by _at least one plugin was..._

Wadeck replied:
In that case that's perfectly fine 👍

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
+     * Print the result of all the process
+     * @param results the list of results for the disablement of each plugin
+     */
+    private void printResults( List<PluginWrapper.PluginDisableResult> results) {
```
nit: `printResults( List` => `printResults(List`

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
+        } else {
+            String f = "%" + indent + "s" + format + "%n";
+            String[] newArgs = new String[arguments.length + 1];
+            newArgs[0] = " "; int i = 1;
```
Please avoir multiple statements per line, we are not writing JavaScript ;)

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
+            String f = "%" + indent + "s" + format + "%n";
+            String[] newArgs = new String[arguments.length + 1];
+            newArgs[0] = " "; int i = 1;
+            for (String arg: arguments ) {
```
nit: `for (String arg: arguments ) {` => `for (String arg: arguments) {` (removal of space)

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
-        return result;
+        printIndented(indent, Messages.DisablePluginCommand_StatusMessage(oneResult.getPlugin(), oneResult.getStatus(), oneResult.getMessage()));
+        if(oneResult.getDependantsDisableStatus().size() > 0) {
+            indent = (indent == 0) ? INDENT_SPACE : indent*2;
```
nit: `indent * 2` (add spaces around operators)

MRamonLeon replied:
BTW, it was an error. It should be _indent + INDENT_SPACE_:
* indent * 2: 3, 6, 12, 24, 48, ...
* indent + INDENT_SPACE: 3, 6, 9, 12, 15, 18, ...

:-D

Wadeck commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -68,68 +78,161 @@ protected int run() throws Exception {
+     */
+    private void printResult(PluginWrapper.PluginDisableResult oneResult, int indent) {
+        PluginWrapper.PluginDisableStatus status = oneResult.getStatus();
+        if(quiet && (PluginWrapper.PluginDisableStatus.DISABLED.equals(status) || PluginWrapper.PluginDisableStatus.ALREADY_DISABLED.equals(status))) {
```
nit: `if(quiet` => `if (quiet`

(also applicable to other occurences, like in `restartIfNecessary`)

Wadeck commented on core/src/main/resources/hudson/cli/Messages.properties:
```diff
@@ -115,10 +115,8 @@ CliProtocol.displayName=Jenkins CLI Protocol/1 (unencrypted)
 DisablePluginCommand.PrintUsageSummary=\
-Disable the plugins with the given short names. Plugins are disabled in order, so make sure to include plugins with dependencies before their dependencies.
+Disable the plugins with the given short names. You can define how to proceed with the dependant plugins and if a restart after should be done. You can also set the quiet mode to avoid extra info in the console.
+DisablePluginCommand.StatusMessage=Disabling {0}: {1} ({2})
```
nit: missing trailing line (same for _es)

Wadeck commented on test/src/test/java/hudson/cli/DisablePluginCommandTest.java:
```diff
@@ -138,6 +182,159 @@ public void disablePluginsWithError16() {
+    @Test
+    @Issue("JENKINS-27177")
+    @WithPlugin({"variant.hpi", "depender-0.0.2.hpi", "mandatory-depender-0.0.2.hpi", "plugin-first.hpi", "dependee-0.0.2.hpi", })
+    public void disablePluginsStrategyAll() {
```
`StrategyAll` please use the same name for the previous tests, the `St` is not really good as abbreviation from my PoV.

MRamonLeon replied:
I used it running away from large method names. What a paradox, because I really like it

Wadeck commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -530,6 +564,95 @@ public void disable() throws IOException {
+        }
+
+        return result;
+    }
```
nit: This method is ~80 lines long, that seems a bit too much. Perhaps you can put some of the logic inside the different strategies (more OOP) to avoid mixing the behavior in this method.

MRamonLeon replied:
It's due to the comments, that I think are necessary. Anyway, I have encapsulated the dependantsToCheck in a method.

Wadeck reviewed (changes requested):
Thank you for the new strategies!

I put some comments about the coding style that are not blocking (nit ones) but other are from my PoV.

The logic seems fine 👍

MRamonLeon committed:
[JENKINS-27177] Minor fixes

MRamonLeon committed:
[JENKINS-27177] Minor fixes

Wadeck reviewed (approved):
:bee:

varyvol commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1829,6 +1840,34 @@ public MetadataCache createCache() {
+
+            if (plugin == null) {
+                PluginWrapper.PluginDisableResult result = new PluginWrapper.PluginDisableResult(pluginName);
+                result.setMessage(Messages.PluginWrapper_NoSuchPlugin(pluginName));
```
It seems weird to create an object and then set some attributes when you already knew those when creating the object. Maybe creating a new constructor or a helper method?

varyvol replied:
But you haven't used the new constructor here?

varyvol commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -530,6 +573,99 @@ public void disable() throws IOException {
+            } else {
+                // As there is no cycles in the plugin dependencies, the recursion shouldn't be infinite. The
+                // strategy used is the same for its dependants plugins
+                result.addDependantDisableStatus(dependantPlugin.disable(strategy));
```
What if this dependant plugin cannot be disabled? Shouldn't that set "aDependantNotDisabled?

MRamonLeon replied:
Right. Even more, we should add this case as a new status for this plugin.

varyvol commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -530,6 +573,99 @@ public void disable() throws IOException {
+
+            // The dependant plugin doesn't exist, add an error to the report
+            if (dependantPlugin == null) {
+                PluginDisableResult dependantStatus = new PluginDisableResult(dependant);
```
Ditto.

varyvol commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -530,6 +573,99 @@ public void disable() throws IOException {
+                // getDependants returns all the dependant plugins, mandatory or optional.
+                dependantsToCheck = this.getDependants();
+                break;
+            case NONE:
```
Why not putting NONE and MANDATORY together?

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        // disable...
+        List<PluginWrapper.PluginDisableResult> results = jenkins.pluginManager.disablePlugins(strategyToUse, pluginNames);
```
Should we check for null or empty lists?

varyvol replied:
You are still not checking if "pluginNames" is null or empty.

MRamonLeon replied:
Added the Nonnull also

varyvol replied:
I am not talking about the annotations, but about checking here if pluginNames == null and then throwing an exception.

MRamonLeon replied:
The parameter is required (`required=true` in the `@Argument` annotation of the class) so it cannot be null. If the user set "" as one of the plugins, the command will result in `NO_SUCH_PLUGIN` for this specific plugin

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        printIndented(indent, Messages.DisablePluginCommand_StatusMessage(oneResult.getPlugin(), oneResult.getStatus(), oneResult.getMessage()));
+        if (oneResult.getDependantsDisableStatus().size() > 0) {
```
Why not the for directly and you can save this if?

varyvol replied:
Why not calling the method inside the for with indent + INDENT_SPACE? You are not using the variable for anything after that.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,239 @@
+            stdout.format(format + "%n", (Object[]) arguments);
+        } else {
+            String f = "%" + indent + "s" + format + "%n";
+            String[] newArgs = new String[arguments.length + 1];
```
I think there needs to be a more comprehensible way of doing this, it seems to complicated just to add a new string in the first position of an array.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,239 @@
+     */
+    private void restartIfNecessary(List<PluginWrapper.PluginDisableResult> results) throws RestartNotSupportedException {
+        if (restart) {
+            for (PluginWrapper.PluginDisableResult oneResult : results) {
```
Maybe a stream() with anymatch()?

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        if (oneResult.getDependantsDisableStatus().size() > 0) {
+            for (PluginWrapper.PluginDisableResult oneDependantResult : oneResult.getDependantsDisableStatus()) {
```
Ditto streams.

varyvol commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -0,0 +1,239 @@
+    private boolean restartIfNecessary(PluginWrapper.PluginDisableResult oneResult) throws RestartNotSupportedException {
+        PluginWrapper.PluginDisableStatus status = oneResult.getStatus();
+        if (PluginWrapper.PluginDisableStatus.DISABLED.equals(status)) {
+            Jenkins.get().safeRestart();
```
Does this wait until the CLI command finishes?

varyvol replied:
I know that, but does it way until the CLI command finishes?

MRamonLeon replied:
It does Jenkins#safeRestart so no, it doesn't

varyvol replied:
I'm not sure if this ideal then, but the "enable-plugin" command works the same, so let's leave it like that.

varyvol commented on test/src/test/java/hudson/cli/DisablePluginCommandTest.java:
```diff
@@ -0,0 +1,376 @@
+        assertPluginDisabled("depender");
+        assertPluginDisabled("plugin-first");
+        assertPluginDisabled("mandatory-depender");
+        assertJenkinsInQuietMode(); // some plugins were disabled
```
I don't understand this comment.

MRamonLeon replied:
changed, i wanted to say that Jenkins is restarting because at least one plugin was disabled.

varyvol commented on test/src/test/java/hudson/cli/DisablePluginCommandTest.java:
```diff
@@ -0,0 +1,376 @@
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.*;
+
+public class DisablePluginCommandTest {
```
I think a test demonstrating that the order in which the plugins are specified does not matter is importante here.

MRamonLeon replied:
There are some tests that demonstrate it, for example: _quietModeEmptyOutputSucceed_, _quietModeWithErrorNoSuch_

varyvol replied:
How are they proving that?

MRamonLeon replied:
no worries, I've created a specific test to demo it: _canDisableDependentPluginWrongOrderStrategyAll_

MRamonLeon committed:
[JENKINS-27177] Add new status error_disabling and minor fixes

When there is some problem disabling a plugin (the disabling file cannot
be created) it should be indicated and also reflected in the dependee
plugin.

MRamonLeon committed:
[JENKINS-27177] Minor fixes.

Change the way of an array copy and ensure non null args

MRamonLeon commented on core/src/main/java/hudson/cli/DisablePluginCommand.java:
```diff
@@ -119,14 +119,10 @@ private void printIndented(int indent, String format, String... arguments) {
-                newArgs[i] = arg;
-                i++;
-            }
+            System.arraycopy(arguments, 0, newArgs, 1, arguments.length);
```
WYT @varyvol ? :wink:

varyvol replied:
Better, but I think you removed too much (the assigment to newArgs[0]).

MRamonLeon committed:
[JENKINS-27177] Improve messages

MRamonLeon committed:
[JENKINS-27177] Specific test to demo right disablement with wrong order

MRamonLeon commented:
@daniel-beck As long as the tests don't fail you can remove the _needs-fix_ and _work-in-progress_ labels.

MRamonLeon commented:
@daniel-beck, you can remove the _needs-fix_ and _work-in-progress labels_.

batmat commented:
I plan to merge this today if nobody objects or does it before. Didn't review it myself, but given there are many approvals and all issues have seemingly been addressed, we can move forward.

batmat commented:
Slightly adjusted the proposed changelog. Merging.
